python/ray/data/_internal/plan.py

```diff
@@ -1163,7 +1163,7 @@ def _reorder_stages(stages: List[Stage]) -> List[Stage]:
             reorder_buf.append(s)
         else:
             # Barrier: flush the reorder buffer.
-            if isinstance(s, AllToAllStage):
+            if isinstance(s, AllToAllStage) or s.name == "write":
```

Comment:
Could we have some interface or attribute about non reorder ability?

Reply:
Yeah, we may do so in new plan since legacy Stage is going to be deprecated (this PR fixes the legacy code path as new plan isn't enabled yet).

Reply:
Are we planning to fix this in the new planner's optimization rule as a follow-up?

Reply:
Yeah, this is fixing an issue that's already breaking.